Upgrade to 2022.3.62f3 LTS

Z-Machine opened the pull request:
This PR supersedes the initial work done by Interkarma on his branch.

This upgrade fixes multiple breaking issues I've experienced under Wayland in the unity2019 editor and also fixes #2728 

Currently everything vanilla seems to work in the editor and in my local build on Linux. I haven't got around to testing mods yet but there doesn't seem to have been any apparent breaking changes.

I'd love to help get this tested and pushed through as I'm sorely lacking newer language features in C# and support for SourceGenerators. :pray:

Z-Machine committed:
Upgrade to latest recommended .gitignore

Z-Machine committed:
Upgrade com.unity.postprocessing and move to registry

Z-Machine committed:
Upgrade packages

Z-Machine committed:
Upgrade to 2022.3.60f1 LTS

Z-Machine committed:
Missing asterisk

Z-Machine committed:
Add package NuGetForUnity

KABoissonneault commented:
I'm most concerned with precompiled mods, which link to specific details of Unity binaries. If we can find some of them and verify that they work without errors, I'd be good with it

Z-Machine committed:
Add package csprojmodifier

Z-Machine committed:
Enable nullable types and upgrade LangVersion to 11

Z-Machine commented:
I've got a local build running with ~110 mods in my current save, it's playable but there are some issues.

It seems Unity has changed how they handle improperly sized DXT5 textures which is causing them to fail to load instead of falling back to RGBA32 which will need to be addressed.

Additionally, around five mods are failing to initialize due to types being incomplete during reflection, and UBLAMF's update checker is failing to compile from source.

Z-Machine committed:
Fix LangVersion.props path

Z-Machine committed:
Add NuGet package Microsoft.CodeAnalysis.CSharp

Z-Machine committed:
Initial compiler replacement

Z-Machine committed:
Forgot to update ModManager to new method

Z-Machine committed:
Remove unneeded import

Z-Machine commented:
Did some initial work over at https://github.com/Z-Machine/daggerfall-unity/pull/1 to replace the runtime compiler with Roslyn. Currently it fixes my above described issue with mods.

Z-Machine committed:
Check if the texture can be compressed

Z-Machine committed:
Check if the texture can be compressed

Z-Machine commented:
![image](https://github.com/user-attachments/assets/545a445e-18e2-4031-993d-26736cc0e02c)
![image](https://github.com/user-attachments/assets/1cabd45e-8a94-43ff-b607-85ffa63388c0)

There are a lot of Texture2Ds being created all over the codebase but the above commit re-implements the fallback behavior.

Z-Machine committed:
Rip out execute console command

Z-Machine committed:
Build pdb at runtime for debugging

Z-Machine committed:
Add pdb support to precompiled mods

Z-Machine committed:
load pdb for precompiled if they exist

Z-Machine committed:
remove rest of old compiler and references

Z-Machine committed:
Switch to Roslyn

Replace runtime compiler with Roslyn and remove old compiler from project.

- Enables nullable types
- LangVersion is now 11
- Mods can now be built in debug mode and the game loads their pdb if found

Z-Machine commented:
> I think we'll need a test build and community testing once you have a build you think is stable enough for general users.

I've got standalone builds published on my fork and have made a forum post, hopefully it generates some interest.

Z-Machine committed:
Fixes runtime crash for multiple mods

Z-Machine commented:
The above commit fixes several mods crashing at runtime as noted by Excoriated for the following mods:

Ambient Text 1.7
Diverse Weapons 1.7
Dungeon Loot 2.0
Famous Face of the Iliac Bay 2.0.1
Readied Spellcasting Hands 0.11.4
Sit and Loiter 1.2
The Red Lantern Guild 2.0
Uncanny UI 1.2 (not jagget fork, jagget fork works)
Villager Reactions 1.2

I am unable to reproduce the crashes under Arch but they says it fixes the crashes for them under Win10.

I've also uploaded a new release over on my fork.

Z-Machine committed:
Revert to old default behavior for physical key mapping

Z-Machine commented:
2022 added a new default input setting that maps physical key locations to key codes per keyboard layout/locale. The above commit reverts this to 2019 behavior until we can switch to the new default setting.

> As of 2022.1 the option is enabled by default with intent of deprecating and removing old behavior in the future.

https://docs.unity3d.com/2022.3/Documentation/Manual/UpgradeGuide2022LTS.html#keys

TwelveEyes commented:
@Z-Machine

Could you embed the `com.unity.postprocessing` package `v3.4.0` in the repo again and restore the changes in this commit: https://github.com/Interkarma/daggerfall-unity/commit/88ace56de6416f89233507967a5d2c8cb67603f2. Looks like it got clobbered in the Unity upgrade.

Z-Machine commented:
> @Z-Machine
> 
> Could you embed the `com.unity.postprocessing` package `v3.4.0` in the PR again and restore the changes in this commit: [88ace56](https://github.com/Interkarma/daggerfall-unity/commit/88ace56de6416f89233507967a5d2c8cb67603f2). Looks like it got clobbered in the Unity upgrade.

Do you actually need it embedded and at that specific version for something? I noticed no issues with any post processing effects in-game or in editor during my testing.

TwelveEyes commented:
>I can revert the commit and roll back to the older version of PPv2 with DFU's patch applied (assuming other packages don't require the newer version) or... pull the newer version into the repo and reapply the patch.

I think the only real option is to bite the bullet and embed PP again. Embedding the older version of PP seems like it goes against the spirit of this PR, so I think embedding the newer version and reapplying the patch is the best option. Unfortunately the extra noise is kinda inevitable.

TwelveEyes commented:
>I've gone ahead and done just that and.. well I can't tell a difference between dithering being enabled or disabled. Furthermore, seemingly the dithering patch has no affect in RetroMode at all.

Hmm, interesting. Main reason I made the comment about dither is a user [posted](https://discord.com/channels/559842137374457876/1205222743000481852/1384485847310794865) on Discord about seeing [this](https://github.com/Interkarma/daggerfall-unity/issues/2246) old issue crop up where dither noise would appear on sprites with posterization/palettization enabled but dithering _disabled_ (Unity's automatic "magic dither").

Daneel53 committed:
Merge branch 'master' into upgrade/2022_3_60f1-lts

Daneel53 committed:
Update ProjectVersion.txt

Upgrade to latest 2022 version 2022.3.62f3

Daneel53 committed:
Cleanup and get rid of warnings

Daneel53 committed:
Move packages and added to .gitignore

Daneel53 committed:
Suppress files into Assets

Daneel53 committed:
Update RmbBlockEditorWindowStyles.uss

Add missing px for font-size

Daneel53 committed:
Update packages

Suppressed Microsoft.CodeAnalysis.Analyzers.3.11.0
Added five installed packages needed.

Daneel53 committed:
Add packages

Daneel53 committed:
Complete jobs before ending schedule in DaggerfallBillboardBatch

Ensure all jobs are completed before proceeding.

Daneel53 committed:
Fix typo in Dependency completion call

Daneel53 committed:
Suppress warning when pdb not found

Daneel53 committed:
Merge remote-tracking branch 'upstream/master' into pr/2733

Daneel53 committed:
Merge branch 'upgrade/2022_3_60f1-lts' of https://github.com/Z-Machine/daggerfall-unity into pr/2733
